rbd: Allow user to remove snapshot with --force to auto flatten children

yangdongsheng opened the pull request.

dillaman commented:
@yangdongsheng What's the use-case for this feature? It seems this is going to offer way to cause mass corruption within the system.

dillaman commented on src/librbd/internal.cc:
```diff
@@ -865,6 +865,62 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
+	  return r;
+	}
+
+	ImageCtx *imctx = new ImageCtx(name, "", NULL, ioctx, false);
```
Open using the known image id and skip the directory lookup

yangdongsheng replied:
right, thanx.

dillaman commented on src/include/rbd/librbd.h:
```diff
@@ -70,6 +70,10 @@ typedef struct {
 #define RBD_MAX_IMAGE_NAME_SIZE 96
 #define RBD_MAX_BLOCK_NAME_SIZE 24
 
+#define RBD_SNAP_REMOVE_UNPROTECT	1 << 1
```
1 << 0?

yangdongsheng replied:
Oh, yes.

dillaman commented:
Without the deep-flatten feature enabled on the clones, if the clones have a single snapshot you will still not be able to unprotect the parent snapshot after flatten.

yangdongsheng commented:
@dillaman This idea comes from http://tracker.ceph.com/issues/6626. I implemented it in librbd here then we can take the use of this feature in CLI and API. 

I have added a remove_child() in flatten_children if needed. Then we can remove snapshot even his children have deep-flatten disabled.

dillaman commented on src/include/rbd/librbd.hpp:
```diff
@@ -247,6 +247,7 @@ class CEPH_RBD_API Image
   int snap_exists2(const char *snapname, bool *exists);
   int snap_create(const char *snapname);
   int snap_remove(const char *snapname);
+  int snap_remove2(const char *snapname, uint32_t flags);
```
It might be nice to add a progress context callback since a flatten will take non-constant time. Same comment for the C++ API.

yangdongsheng replied:
okey.

dillaman commented on src/test/cli/rbd/help.t:
```diff
@@ -1224,6 +1225,7 @@
     -p [ --pool ] arg    pool name
     --image arg          image name
     --snap arg           snapshot name
+    --force              force to remove snapshot even it's protected
```
Minor: should also indicate that it will flatten any associated clones.

yangdongsheng replied:
good suggestion

dillaman commented on src/tools/rbd/action/Snap.cc:
```diff
@@ -255,13 +260,13 @@ int execute_remove(const po::variables_map &vm) {
   librados::IoCtx io_ctx;
   librbd::Image image;
-  r = utils::init_and_open_image(pool_name, image_name, "", false, &rados,
+  r = utils::init_and_open_image(pool_name, image_name, snap_name, false, &rados,
```
Why was this required?

yangdongsheng replied:
The all operations in tools/rbd/Snap.cc are working on snapshot. So I think we can get a context for a specified snapshot. In this case, I need to get imctx.snap_id later.
```
parent_spec parent_spec(ictx->md_ctx.get_id(), ictx->id, ictx->snap_id);
```
https://github.com/ceph/ceph/pull/10087/files#diff-254555dde8dcfb7fb908791ab8214b92R874

yangdongsheng replied:
But I can get the snap_id by ```get_snap_id(snap_name);``` in this case. I will update it.

dillaman commented on src/librbd/internal.cc:
```diff
@@ -2248,6 +2310,41 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
+    if (r < 0 && r != -EBUSY)
+      return r;
+   
+    if (r == -EBUSY && (flags & RBD_SNAP_REMOVE_UNPROTECT)) {
```
Why not just test for needs unprotection or needs children flattening first instead of requiring a goto.

yangdongsheng replied:
Oh, yes. I will remove the goto.

yangdongsheng commented:
@dillaman updated.

dillaman commented:
@yangdongsheng Can you add one or more test cases to test_librbd to exercise the new API?

yangdongsheng commented:
@dillaman Sure, test case added.

dillaman commented on src/tracing/librbd.tp:
```diff
@@ -1089,6 +1089,32 @@ TRACEPOINT_EVENT(librbd, snap_remove_exit,
+        const char*, snap_name,
+        char, read_only,
+        const char*, remove_snap_name,
+	uint32_t flags),
```
Missing the comma -- fails to compile w/ LTTng enabled.

yangdongsheng commented:
@dillaman Fixed. sorry for the foolish mistake.

dillaman commented:
Looks like the cram test needs to be updated to handle the new "Removing snap" output from 'rbd snap rm':

http://teuthology.ovh.sepia.ceph.com/teuthology/jdillaman-2016-07-22_00:37:26-rbd-wip-jd-testing---basic-openstack/41129/teuthology.log

yangdongsheng commented:
@dillaman Yes, I added a --no-progress in format.t now. Please help to queue this into integration again. Thanx a lot.

yangdongsheng commented:
Hi @dillaman and @trociny I found a fail in http://pulpito.ceph.com/?date=2016-07-26&branch=wip-jd-testing
```
2016-07-26T07:25:36.987 INFO:tasks.cram.client.0.vpm079.stdout: # cleanup
2016-07-26T07:25:36.987 INFO:tasks.cram.client.0.vpm079.stdout:   $ rbd snap remove rbd_other/deep-flatten-child@snap
2016-07-26T07:25:36.988 INFO:tasks.cram.client.0.vpm079.stdout:+
2016-07-26T07:25:36.988 INFO:tasks.cram.client.0.vpm079.stdout:+  Removing snap: 100% complete...done.
2016-07-26T07:25:36.988 INFO:tasks.cram.client.0.vpm079.stdout:   $ rbd snap remove rbd_other/child@snap
2016-07-26T07:25:36.988 INFO:tasks.cram.client.0.vpm079.stdout:+
2016-07-26T07:25:36.989 INFO:tasks.cram.client.0.vpm079.stdout:+  Removing snap: 100% complete...done.
2016-07-26T07:25:36.989 INFO:tasks.cram.client.0.vpm079.stdout:   $ rbd snap unprotect bar@snap
```

But I checked the branch of wip-jason-testing, the format.t is really really updated and different with above.

I am not sure how to read the result of pulpito.ceph.com, but I found it's strange the "update time" is later than "start time" and "scheduled time". Should I wait for another schedule for the latest update?

dillaman commented:
@yangdongsheng Some tests unfortunately always pull from the master branch -- this is one of those tests

yangdongsheng commented:
@dillaman Oh I see it . it's from "wget -nc -nv -P /home/ubuntu/cephtest/archive/cram.client.0 -- 'http://git.ceph.com/?p=ceph.git;a=blob_plain;f=src/test/cli-integration/rbd/formatted-output.t'". thanx for your explanation.
